ebu_tt_live/node/deduplicator.py

```diff
@@ -0,0 +1,199 @@
+
+
+class DeDuplicatorNode(AbstractCombinedNode):
+    _sequence_identifier = None
+    _expects = EBUTT3Document
+    _provides = EBUTT3Document
+
+    def __init__(self, node_id, sequence_identifier, consumer_carriage=None, \
+                 producer_carriage=None):
+        super(DeDuplicatorNode, self).__init__(
+            node_id=node_id,
+            consumer_carriage=consumer_carriage,
+            producer_carriage=producer_carriage
+        )
+        self._sequence_identifier = sequence_identifier
+
+    def process_document(self, document, **kwargs):
+        if self.is_document(document):
+
+            if document.sequence_identifier == self._sequence_identifier:
+                raise UnexpectedSequenceIdentifierError()
+
+            if self.check_if_document_seen(document=document):
+
+                self.limit_sequence_to_one(document)
+
+                document.sequence_identifier = self._sequence_identifier
+
+                self.remove_duplication(document=document)
+                print document.get_xml()
+
+                document.validate()
+                self.producer_carriage.emit_data(data=document, **kwargs)
+
+    def remove_duplication(self, document):
+        _old_id_dict = dict({})
+        _new_id_dict = dict({})
+        hash_dict = dict({})
+
+        if document.binding.head.styling is not None and document.binding.head.layout is not None:
+            styles = document.binding.head.styling.style
+            document.binding.head.styling.style = None
+
+            regions = document.binding.head.layout.region
+            document.binding.head.layout.region = None
+
+            self.CollateUniqueVals(styles, _old_id_dict, _new_id_dict, hash_dict)
+            self.AppendNewElements(styles, document.binding.head.styling.style, \
+                                   _old_id_dict, _new_id_dict, hash_dict)
+
+            self.CollateUniqueVals(regions, _old_id_dict, _new_id_dict, hash_dict)
+            self.AppendNewElements(regions, document.binding.head.layout.region, \
+                                   _old_id_dict, _new_id_dict, hash_dict)
+
+            replace_id_refs = ReplaceStylesAndRegions(document.binding, \
+                                                      _old_id_dict, \
+                                                      _new_id_dict)
+            replace_id_refs.proceed()
+        else:
+            return None
+
+    def CollateUniqueVals(self, element_list, _old_id_dict, _new_id_dict, \
+                          hash_dict):
+        for value in element_list:
+            #deduplicating in-line styles
+            if value.style is not None:
+                for old_id_index in range(len(value.style)):
+                    old_id_ref = _old_id_dict.get(value.style[old_id_index])
+                    new_id_ref = _new_id_dict.get(old_id_ref)
+
+                    value.style[old_id_index] = new_id_ref
+
+            #deduplicating region elements
```

Comment:
I think there's a region element you've hurt the feelings of.

(I've amended the comment)